### ISSUE TITLE
Global pollution

### ISSUE THREAD

**Issue by GordonSmith** [OpenNTF/dojo-webpack-plugin-sample#54]
I am running into a few issues with global pollution.

My goal is to wrap dgrid into a single JS UMD library file, with only a few items exported, which can:
* Get AMD loaded via some unknown AMD loader (requirejs, dojo, d3-require etc.)
* Get bundled into a web application (webpack, rollupjs, ruiquirejs build etc)

Issues:
1. When loading via AMD, any existing global "require" function will get overwritten by the dojo one.
2. When re-bundled in webpack there is no global "this" scope, causing errors on initial load.
3. When re-bundled in rollupjs a small portion of the dojo loader code ends up inside a "use strict;" scope which causes it to fail.

For 1 and 2 there is a relatively trivial fix, but will need some discussion as I _think_ your unit tests rely on the webpacked bundle to inject a global require.

For 3, I need to do some additional research as there are several places the "use strict" gets injected (tsc compiler and webpack - but only when it thinks it is a es6 module).

(See https://github.com/OpenNTF/dojo-webpack-plugin/pull/92 for my "quick" fix for 1+2)

**Comment by chuckdumont** [comment]
Some info on unit tests that might help clarify things.  The test runner, test/test.js, passes the node require function to the application under test as the first parameter, so require is defined with this value in the top level context of the application.  See https://github.com/OpenNTF/dojo-webpack-plugin/blob/v2.2.1/test/test.js#L134.  Then code in test/plugins/MainTemplatePlugin modifies the MainTemplatePlugin `require` and `render` events to add a module scoped require for AMD modules that reference the Dojo implementation.

```javascript
compilation.mainTemplate.plugin("require", function(source) {
	// When running unit tests in node, the scoped require function is node's CommonJs
	// require.  This code, together with the 'render' plugin below, defines a scoped
	// Dojo require variable for each AMD module so that referencing 'require' from
	// within the module will yield the Dojo function.
	return source.replace(/__webpack_require__\);/g, "__webpack_require__, req);");
});

compilation.moduleTemplate.plugin("render", function(source, module) {
	var result = source;
	if (module.isAMD) {
		// Define a module scoped 'require' variable for AMD modules that yields the
		// the Dojo require function.
		result = new ConcatSource();
		result.add("var require = arguments[3];");
		result.add(source);
	}
	return result;
});
```

This allows bootstrap code to call node's require to load async chunks, but when require is called from within an AMD module, the plugin's implementation of Dojo require will be called.

It wasn't thought that the module scoped require was needed in the web application because require is defined with the correct Dojo value in global scope, but if overwriting the global require is an issue, then perhaps this approach should be used for application code as well as unit tests.

**Comment by chuckdumont** [comment]
Also, I think that the behavior to **not** define global require should be controlled by an option, with the default being to define it.  This would make the default behavior consistent with the Dojo behavior.

**Comment by chuckdumont** [comment]
@GordonSmith, does the solution referenced in the PR resolve this issue.  Can this be closed now?

**Comment by GordonSmith** [comment]
Apologies I have not had a chance to retest (yet), but will report back as soon as I have.

**Comment by chuckdumont** [comment]
Closing as fixed in 2.3.0.  You can re-open if you still have problems
